src/crimson/osd/pg.cc

```diff
@@ -610,6 +610,8 @@ void PG::init(
   peering_state.init(
     role, newup, new_up_primary, newacting,
     new_acting_primary, history, pi, t);
+  assert(coll_ref);
+  t.touch(coll_ref->get_cid(), make_snapmapper_oid());
```

Comment:
Why is this section not part of `OSD::start()` instead?

Reply:
We should also check for existence and add logs if possible.
```
// make sure snap mapper object exists
if (!store->exists(coll_ref, make_snapmapper_oid())) {
  logger().debug("init creating/touching snapmapper object");
  ..
  t.touch(coll_ref->get_cid(), make_snapmapper_oid());
}
```

Reply:
Since, in crimson, we previously store snap mapper in pgmeta objects which are pg specific, I made the snap mapper objects pg specific, too. I think you are suggesting that we make the snapmapper object shared across the whole OSD like what classic OSD is doing. Am I right?

Reply:
Hmm.  The reason classic places all the snap trimmer information in a single object is that otherwise we'd need to read and split/merge the objects during pg split/merge.  This would require IO linear in the number of snapshots, and split/merge aren't otherwise dependent on the size of the PGs (for good reason -- they must occur during the map transition and block IO/peering on those PGs).  Crimson will have different problems for split/merge -- specifically that seastore will want to maintain per-shard metadata trees, so we won't be able to merge PGs easily on different shards or split into a different shard.  For now, I'd actually suggest a single OSD-wide object maintained by shard 0 (pgs on other shards will have to proxy IO to that shard).  If this becomes a bottleneck, we can try something else later.

Reply:
Done.

Reply:
> Hmm. The reason classic places all the snap trimmer information in a single object is that otherwise we'd need to read and split/merge the objects during pg split/merge. This would require IO linear in the number of snapshots, and split/merge aren't otherwise dependent on the size of the PGs (for good reason -- they must occur during the map transition and block IO/peering on those PGs). Crimson will have different problems for split/merge -- specifically that seastore will want to maintain per-shard metadata trees, so we won't be able to merge PGs easily on different shards or split into a different shard. For now, I'd actually suggest a single OSD-wide object maintained by shard 0 (pgs on other shards will have to proxy IO to that shard). If this becomes a bottleneck, we can try something else later.

@athanatos Hi, sam. I think the modification to the OSD-wide object and other modifications have to be in the same transaction, am I right? If so, it looks to me that we need to synchronize transactions across multiple shards if they all modify the OSD-wide object, am I right?

Reply:
Yeah, I forgot that the sharded objectstore abstraction doesn't allow transactions that span shards.  A per-pg object seems fine, we'll figure out pg split/merge later.

Reply:
> Yeah, I forgot that the sharded objectstore abstraction doesn't allow transactions that span shards. A per-pg object seems fine, we'll figure out pg split/merge later.

Done:-)